ethcore/src/ethereum/ethash.rs

```diff
@@ -437,13 +441,28 @@ impl Engine for Arc<Ethash> {
 		Ok(())
 	}
 
+	fn verify_transaction(&self, t: UnverifiedTransaction, header: &Header) -> Result<SignedTransaction, Error> {
+		let signed = SignedTransaction::new(t)?;
+		if !self.tx_filter.as_ref().map_or(true, |filter| filter.transaction_allowed(header.parent_hash(), &signed)) {
+			return Err(From::from(TransactionError::NotAllowed));
+		}
+		Ok(signed)
+	}
+
 	fn epoch_verifier<'a>(&self, _header: &Header, _proof: &'a [u8]) -> engines::ConstructedVerifier<'a> {
 		engines::ConstructedVerifier::Trusted(Box::new(self.clone()))
 	}
 
 	fn snapshot_components(&self) -> Option<Box<::snapshot::SnapshotComponents>> {
 		Some(Box::new(::snapshot::PowSnapshot::new(SNAPSHOT_BLOCKS, MAX_SNAPSHOT_BLOCKS)))
 	}
+
+	fn register_client(&self, client: Weak<Client>) {
+		if let Some(ref filter) = self.tx_filter {
+			filter.register_client(client as Weak<BlockChainClient>);
```

Comment:
engines are used in light clients too. requiring a full node to run consensus logic should have been a blocker for this PR.

Reply:
furthermore, I don't see why TX filtering would be coupled with the Ethash memory-hard PoW consensus system in the first place.

Reply:
it makes more sense in a permissioned environment anyway